app/models/concerns/unparsed_run.rb

```diff
@@ -4,163 +4,290 @@ module UnparsedRun
+              filename: "#{id}_#{i}"
+            )
+          end
+
+          segment.realtime_end_ms = lsc_segment.personal_best_split_time.real_time&.total_seconds&.*(1000) || 0
+          segment.realtime_duration_ms = segment.realtime_end_ms - previous_segment.realtime_end_ms
+          segment.realtime_start_ms = segment.realtime_end_ms - segment.realtime_duration_ms
+          segment.realtime_shortest_duration_ms = lsc_segment.best_segment_time.real_time&.total_seconds&.*(1000) || 0
+          segment.realtime_skipped = segment.realtime_duration_ms.zero?
+          segment.realtime_gold = segment.realtime_duration_ms <= segment.realtime_shortest_duration_ms
+          segment.realtime_reduced = false
+
+          segment.gametime_end_ms = lsc_segment.personal_best_split_time.game_time&.total_seconds&.*(1000) || 0
+          segment.gametime_duration_ms = segment.gametime_end_ms - previous_segment.gametime_end_ms
+          segment.gametime_start_ms = segment.gametime_end_ms - segment.gametime_duration_ms
+          segment.gametime_shortest_duration_ms = lsc_segment.best_segment_time.game_time&.total_seconds&.*(1000) || 0
+          segment.gametime_skipped = segment.gametime_duration_ms.zero?
+          segment.gametime_gold = segment.gametime_duration_ms <= segment.gametime_shortest_duration_ms
+          segment.gametime_reduced = false
+
+          container[lsc_segment] = segment
+          previous_segment = segment
+        end
 
-      segs.each.with_index do |seg, i|
-        return nil if seg[:history].blank?
+        result = write_segments(container.values)
+        process_segment_histories(container.keys, result.ids)
+        container = {}
+      end
+    end
+
+    def write_segments(segments)
+      # nil out all blank times here since we needed them before to be 0's for finding times
+      segments.each do |seg|
+        %w[real game].each do |timing|
+          %w[time_end_ms time_duration_ms time_shortest_duration_ms].each do |field|
+            value = seg.send("#{timing}#{field}")
+            seg.send("#{timing}#{field}=", zero_to_nil(value))
+          end
+        end
+      end
 
-        seg[:history].each do |history|
-          next unless history[:attempt_number].to_i.positive?
+      result = Segment.import(segments)
+      raise UnparsableRun if result.failed_instances.any?
 
-          histories << SegmentHistory.new(
-            segment_id:           ids[i],
-            attempt_number:       history[:attempt_number].to_i,
-            realtime_duration_ms: history[:realtime_duration_ms],
-            gametime_duration_ms: history[:gametime_duration_ms]
-          )
+      segments.each do |segment|
+        segment.save
+      end
+      result
+    end
+
+    def process_segment_histories(lsc_segments, segment_ids)
+      container = []
+
+      lsc_segments.each_with_index do |seg, i|
+        seg.segment_history.iter.with do |iter|
+          until (history_element = iter.next).nil?
```

Comment:
I wasn't sure how to use `each_slice` with this one.  Any idea/other ideas to clean this one up?

Reply:
Ah with these weird unidiomatic iterators I wouldn't worry about it too much.